src/sql/src/query_model/rewrite/rule/simplify_outer_joins.rs

```diff
@@ -0,0 +1,186 @@
+//! ](https://www.academia.edu/26160408/Outerjoin_simplification_and_reordering_for_query_optimization)
+
+use std::collections::HashSet;
+
+use crate::query_model::attribute::core::Attribute;
+use crate::query_model::attribute::propagated_nulls::PropagatedNulls;
+use crate::query_model::attribute::rejected_nulls::RejectedNulls;
+use crate::query_model::model::BoxType;
+use crate::query_model::rewrite::ApplyStrategy;
+use crate::query_model::rewrite::Rule;
+use crate::query_model::rewrite::VisitOrder;
+use crate::query_model::BoxId;
+use crate::query_model::QuantifierId;
+
+pub(crate) struct SimplifyOuterJoins();
+
+/// A struct that indicates an outerjoin simplification action.
+enum Action {
+    /// Simplify a full outer join to a left/right outer join by
+    /// changing the type of [`QuantifierId`] from
+    /// [`QuantifierType::PreservedForeach`] to
+    /// [`QuantifierType::Foreach`].
+    OuterToOuter(BoxId, QuantifierId),
+    /// Simplify a full outer join to an inner join by
+    /// changing the type all quantifiers from
+    /// [`QuantifierType::PreservedForeach`]
+    /// to [`QuantifierType::Foreach`]
+    /// and changing the box type from
+    /// [`BoxType::OuterJoin`] to [`BoxType::Select`].
+    OuterToInner(BoxId),
+}
+
+impl Rule for SimplifyOuterJoins {
+    /// A (non-empty) sequence of simplification actions to be performed.
+    type Match = Vec<Action>;
+
+    fn name(&self) -> &'static str {
+        "SimplifyOuterJoins"
+    }
+
+    fn strategy(&self) -> ApplyStrategy {
+        ApplyStrategy::AllBoxes(VisitOrder::Pre)
+    }
+
+    fn required_attributes(&self) -> std::collections::HashSet<Box<dyn Attribute>> {
+        HashSet::from([
+            Box::new(PropagatedNulls) as Box<dyn Attribute>,
+            Box::new(RejectedNulls) as Box<dyn Attribute>,
+        ])
+    }
+
+    fn check(
+        &self,
+        model: &crate::query_model::Model,
+        box_id: crate::query_model::BoxId,
+    ) -> Option<Self::Match> {
+        // get the rejected nulls at the root
+        let rej_nulls = model.get_box(box_id).attributes.get::<RejectedNulls>();
+
+        // we can return immediately if there are no rejected nulls at this node
+        if rej_nulls.is_empty() {
+            return None;
+        }
+
+        // An optional match found at the current box.
+        let mut mat = vec![] as Self::Match;
+
+        // A stack that keeps track of the number of randing quantifiers
+        // associated with each box along the recursion trace.
+        let mut quantifier_counts = vec![] as Vec<usize>;
+
+        // A stack of the rejected nulls for the predicates associated
+        // with the current box (identified by box_id) expressed in
+        // terms of the input quantifiers of the current box
+        // initialize with the nulls at the root.
+        let mut rej_nulls = vec![rej_nulls.clone()];
+
+        // TODO: this needs an immutable visit method
```

Comment:
@wangandi I think we need to add a non-mutable visit method in order to make the borrow checker happy here.